File: m26-py/m26/distance.py
__author__ = 'cjoakim'

from .constants import Constants


class Distance(object):

    def __init__(self, dist=0.0, uom=Constants.uom_miles()):
        self.value = float(dist)
        self.uom = self.unit_of_measure(uom)

    def unit_of_measure(self, s):
        u = str(s).strip().lower()
        if u == 'k':
            return Constants.uom_kilometers()
        elif u == 'y':
            return Constants.uom_yards()
        else:
            return Constants.uom_miles()

    def is_miles(self):
        return self.uom == Constants.uom_miles()

    def is_kilometers(self):
        return self.uom == Constants.uom_kilometers()

    def is_yards(self):
        return self.uom == Constants.uom_yards()

    def as_miles(self):
        if self.is_miles():
            return self.value
        elif self.is_kilometers():
            return self.value / Constants.kilometers_per_mile()
        else:
            return self.value / Constants.yards_per_mile()

    def as_kilometers(self):
        if self.is_miles():
            return self.value * Constants.kilometers_per_mile()
        elif self.is_kilometers():
            return self.value
        else:
            return self.value / Constants.yards_per_kilometer()

    def as_yards(self):
        if self.is_miles():
            return self.value * Constants.yards_per_mile()
        elif self.is_kilometers():
            return self.value * Constants.yards_per_kilometer()
        else:
            return self.value

    def add(self, another_instance):
        if self.is_miles():
            self.value = self.value + another_instance.as_miles()
        elif self.is_kilometers():
            self.value = self.value + another_instance.as_kilometers()
        else:
            self.value = self.value + another_instance.as_yards()

    def subtract(self, another_instance):
        if self.is_miles():
            self.value = self.value - another_instance.as_miles()
        elif self.is_kilometers():
            self.value = self.value - another_instance.as_kilometers()
        else:
            self.value = self.value - another_instance.as_yards()

    def __str__(self):
        return "<Distance value:{0} uom:{1}>".format(self.value, self.uom)

File: m26-cs/M26/Joakimsoftware.M26/src/Distance.cs
using System;

// Instances of this class represent a Distance of an event, such as a run or swim or cycling.
// The distance can be expressed in miles, kilometers, or yards.
// Chris Joakim, 2021/07/19

namespace Joakimsoftware.M26 {
    public class Distance {
        public double value { get; }
        public string uom { get; }

        public Distance(double val, string unitOfMeasure=Constants.UomMiles) {

            value = val;

            if (string.IsNullOrEmpty(unitOfMeasure)) {
                uom = Constants.UomMiles;  // default to miles
            }
            else {
                if (string.Equals(unitOfMeasure.Trim(), Constants.UomKilometers, StringComparison.OrdinalIgnoreCase)) {
                    uom = Constants.UomKilometers;
                }
                else if (string.Equals(unitOfMeasure.Trim(), Constants.UomYards, StringComparison.OrdinalIgnoreCase)) {
                    uom = Constants.UomYards;
                }
                else {
                    uom = Constants.UomMiles;
                }
            }
        }

        public Boolean isMiles() {

            return uom.Equals(Constants.UomMiles);
        }

        public Boolean isKilometers() {

            return uom.Equals(Constants.UomKilometers);
        }

        public Boolean isYards() {

            return uom.Equals(Constants.UomYards);
        }

        public double asMiles() {

            if (isKilometers()) {
                return value / Constants.KilometersPerMile;
            }
            else if (isYards()) {
                return value / Constants.YardsPerMile;
            }
            else {
                return value;
            }
        }

        public double asKilometers() {

            if (isKilometers()) {
                return value;
            }
            else if (isYards()) {
                return value / Constants.YardsPerKilometer;
            }
            else {
                return value * Constants.KilometersPerMile;
            }
        }

        public double asYards() {

            if (isKilometers()) {
                return value * Constants.YardsPerKilometer;
            }
            else if (isYards()) {
                return value;
            }
            else {
                return value * Constants.YardsPerMile;
            }
        }

        public Distance add(Distance d2) {

            if (d2 == null) {
                return new Distance(value, uom);
            }

            if (isKilometers()) {
                double v = value + d2.asKilometers();
                return new Distance(v, uom);
            }
            else if (isYards()) {
                double v = value + d2.asYards();
                return new Distance(v, uom);
            }
            else {
                double v = value + d2.asMiles();
                return new Distance(v, uom);
            }
        }

        public Distance subtract(Distance d2) {

            if (d2 == null) {
                return new Distance(value, uom);
            }

            if (isKilometers()) {
                double v = value - d2.asKilometers();
                return new Distance(v, uom);
            }
            else if (isYards()) {
                double v = value - d2.asYards();
                return new Distance(v, uom);
            }
            else {
                double v = value - d2.asMiles();
                return new Distance(v, uom);
            }
        }
    }
}
